File: server/Services/FactionInfluenceService.cs
using ColonyTracker.Api.Models;

namespace ColonyTracker.Api.Services;

public interface IFactionInfluenceService
{
    double CalculateCharacterStrength(IEnumerable<Character> members, int additionalMemberCount = 0);
    int CalculateOrganization(Faction faction, IEnumerable<Character> members);
    int CalculateLegitimacy(Faction faction);
    int CalculateEffectivePower(Faction faction, IEnumerable<Character> members);
}

public class FactionInfluenceService : IFactionInfluenceService
{
    private const double AdditionalMemberBaseInfluence = 30.0;

    // CharacterStrength = (avg * 0.6) + (max * 0.4); leaderless penalty applied here.
    // additionalMemberCount pads the pool with baseline-strength generics so large factions
    // don't require fully populating every member to model their mass accurately.
    public double CalculateCharacterStrength(IEnumerable<Character> members, int additionalMemberCount = 0)
    {
        var list = members.ToList();
        var totalCount = list.Count + additionalMemberCount;
        if (totalCount == 0) return 0;

        var representedSum = list.Sum(c => (double)c.Influence);
        var additionalSum  = additionalMemberCount * AdditionalMemberBaseInfluence;
        var avg = (representedSum + additionalSum) / totalCount;
        var max = list.Count > 0 ? list.Max(c => (double)c.Influence) : AdditionalMemberBaseInfluence;
        var raw = avg * 0.6 + max * 0.4;

        var hasLeader = list.Any(c => c.CharacterType == CharacterType.FactionLeader);
        return raw * (hasLeader ? 1.0 : 0.75);
    }

    // Organization = CharacterStrength×0.40 + NormalizedMomentum×0.20
    // (asset weights omitted server-side — assets are scored client-side)
    // normalizedMomentum = 50 + (momentum / 2)
    public int CalculateOrganization(Faction faction, IEnumerable<Character> members)
    {
        var charStrength       = CalculateCharacterStrength(members, faction.AdditionalMemberCount);
        var normalizedMomentum = 50.0 + faction.Momentum / 2.0;

        var raw =
            charStrength       * 0.40
            + normalizedMomentum * 0.20;

        return Math.Clamp((int)Math.Round(raw), 0, 100);
    }

    // Legitimacy = BaseLegitimacy×0.70
    // (asset legitimacy omitted server-side — assets scored client-side)
    public int CalculateLegitimacy(Faction faction)
    {
        var raw = faction.BaseLegitimacy * 0.70;
        return Math.Clamp((int)Math.Round(raw), 0, 100);
    }

    // Power = Organization × (0.5 + Legitimacy / 100) + PowerModifier
    public int CalculateEffectivePower(Faction faction, IEnumerable<Character> members)
    {
        var organization = CalculateOrganization(faction, members);
        var legitimacy   = CalculateLegitimacy(faction);
        var legitimacyMultiplier = 0.5 + legitimacy / 100.0;
        return (int)Math.Round(organization * legitimacyMultiplier + faction.PowerModifier);
    }
}

File: client/src/app/core/services/faction-influence.service.ts
import { Injectable } from '@angular/core';
import { Character, Faction, FormulasConfig, DEFAULT_FORMULAS, Asset, computeAssetInfluence, computeAssetLegitimacy } from '../models/types';

@Injectable({ providedIn: 'root' })
export class FactionInfluenceService {

  // CharacterStrength = (avg * memberAvgWeight) + (max * memberMaxWeight)
  // Leaderless penalty applied here rather than to final power.
  calculateCharacterStrength(members: Character[], f: FormulasConfig = DEFAULT_FORMULAS): number {
    if (members.length === 0) return 0;
    const avg = members.reduce((s, c) => s + c.influence, 0) / members.length;
    const max = Math.max(...members.map(c => c.influence));
    const raw = avg * f.memberAvgWeight + max * f.memberMaxWeight;
    const hasLeader = members.some(c => c.characterType === 'FactionLeader');
    return raw * (hasLeader ? 1 : (f.leaderlessPowerMultiplier ?? 0.75));
  }

  // Normalized asset influence score: raw sum → 0–100
  calculateAssetInfluenceScore(assets: Asset[], factionId: string, f: FormulasConfig = DEFAULT_FORMULAS): number {
    const raw = assets
      .filter(a => a.controllingFactionId === factionId)
      .reduce((sum, a) => sum + computeAssetInfluence(a), 0);
    return Math.min(100, raw / (f.assetInfluenceScale ?? 25) * 100);
  }

  // Normalized asset legitimacy score: raw sum → 0–100
  calculateAssetLegitimacyScore(assets: Asset[], factionId: string, f: FormulasConfig = DEFAULT_FORMULAS): number {
    const raw = assets
      .filter(a => a.controllingFactionId === factionId)
      .reduce((sum, a) => sum + computeAssetLegitimacy(a), 0);
    return Math.min(100, raw / (f.assetLegitimacyScale ?? 25) * 100);
  }

  // normalizedMomentum maps -100..+100 → 0..100
  calculateNormalizedMomentum(momentum: number): number {
    return 50 + momentum / 2;
  }

  // Organization = CharacterStrength×charWeight + AssetInfluenceScore×assetWeight + NormalizedMomentum×momentumWeight
  calculateOrganization(faction: Faction, members: Character[], f: FormulasConfig = DEFAULT_FORMULAS, assets: Asset[] = []): number {
    const charStrength = this.calculateCharacterStrength(members, f);
    const assetInfluenceScore = this.calculateAssetInfluenceScore(assets, faction.id, f);
    const normalizedMomentum = this.calculateNormalizedMomentum(faction.momentum);
    const raw =
      charStrength       * f.charInfluenceWeight
      + assetInfluenceScore * f.assetInfluenceWeight
      + normalizedMomentum  * f.momentumInfluenceWeight;
    return Math.min(100, Math.max(0, Math.round(raw)));
  }

  // Legitimacy = BaseLegitimacy×baseWeight + AssetLegitimacyScore×assetWeight
  calculateLegitimacy(faction: Faction, f: FormulasConfig = DEFAULT_FORMULAS, assets: Asset[] = []): number {
    const assetLegitimacyScore = this.calculateAssetLegitimacyScore(assets, faction.id, f);
    const raw =
      faction.baseLegitimacy  * (f.baseLegitimacyWeight ?? 0.70)
      + assetLegitimacyScore  * (f.assetLegitimacyWeight ?? 0.30);
    return Math.min(100, Math.max(0, Math.round(raw)));
  }

  // Power = Organization × LegitimacyMultiplier + PowerModifier
  calculateEffectivePower(faction: Faction, members: Character[], f: FormulasConfig = DEFAULT_FORMULAS, assets: Asset[] = []): number {
    const organization = this.calculateOrganization(faction, members, f, assets);
    const legitimacy   = this.calculateLegitimacy(faction, f, assets);
    const legitimacyMultiplier = f.legitimacyBase + legitimacy / f.legitimacyScale;
    return Math.round(organization * legitimacyMultiplier + faction.powerModifier);
  }

  // Kept for backwards compat with the faction detail influence card
  calculateCharacterInfluence(members: Character[], f: FormulasConfig = DEFAULT_FORMULAS): number {
    return this.calculateCharacterStrength(members, f);
  }

  // Kept for backwards compat — now Organization is the equivalent of totalInfluence
  calculateTotalInfluence(faction: Faction, members: Character[], f: FormulasConfig = DEFAULT_FORMULAS, assets: Asset[] = []): number {
    return this.calculateOrganization(faction, members, f, assets);
  }

  topMembers(members: Character[], count = 5): Character[] {
    return [...members].sort((a, b) => b.influence - a.influence).slice(0, count);
  }
}
